sys/kern/src/arch/riscv64.rs

```diff
@@ -0,0 +1,603 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Architecture support for RISC-V.
+//!
+//! Interrupts (other than the Machine Timer used to advance the kernel
+//! timestamp) are not yet supported.
+
+use core::arch::asm;
+use core::ptr::NonNull;
+use core::sync::atomic::{AtomicBool, Ordering};
+
+use zerocopy::FromBytes;
+
+use crate::startup::with_task_table;
+use crate::task;
+use crate::time::Timestamp;
+use crate::umem::USlice;
+use abi::{FaultInfo, FaultSource};
+use unwrap_lite::UnwrapLite;
+
+extern crate riscv_rt;
+
+use riscv::register;
+use riscv::register::mcause::{Exception, Interrupt, Trap};
+use riscv::register::mstatus::MPP;
+
+macro_rules! uassert {
+    ($cond : expr) => {
+        if !$cond {
+            panic!("Assertion failed!");
+        }
+    };
+}
+
+/// On RISC-V we use a global to record the current task pointer.  It may be
+/// possible to use the mscratch register instead.
+#[no_mangle]
+static mut CURRENT_TASK_PTR: Option<NonNull<task::Task>> = None;
+
+/// To allow our clock frequency to be easily determined from a debugger, we
+/// store it in memory.
+#[no_mangle]
+static mut CLOCK_FREQ_KHZ: u32 = 0;
+
+/// RISC-V volatile registers that must be saved across context switches.
+#[repr(C)]
+#[derive(Clone, Debug, Default, FromBytes)]
+pub struct SavedState {
+    // NOTE: the following fields must be kept contiguous!
+    ra: u64,
+    sp: u64,
+    gp: u64,
+    tp: u64,
+    t0: u64,
+    t1: u64,
+    t2: u64,
+    s0: u64,
+    s1: u64,
+    a0: u64,
+    a1: u64,
+    a2: u64,
+    a3: u64,
+    a4: u64,
+    a5: u64,
+    a6: u64,
+    a7: u64,
+    s2: u64,
+    s3: u64,
+    s4: u64,
+    s5: u64,
+    s6: u64,
+    s7: u64,
+    s8: u64,
+    s9: u64,
+    s10: u64,
+    s11: u64,
+    t3: u64,
+    t4: u64,
+    t5: u64,
+    t6: u64,
+    // Additional save value for task program counter
+    pc: u64,
+    // NOTE: the above fields must be kept contiguous!
+}
+
+/// Map the volatile registers to (architecture-independent) syscall argument
+/// and return slots.
+impl task::ArchState for SavedState {
+    fn stack_pointer(&self) -> usize {
+        self.sp as usize
+    }
+
+    /// Reads syscall argument register 0.
+    fn arg0(&self) -> usize {
+        self.a0 as usize
+    }
+    fn arg1(&self) -> usize {
+        self.a1 as usize
+    }
+    fn arg2(&self) -> usize {
+        self.a2 as usize
+    }
+    fn arg3(&self) -> usize {
+        self.a3 as usize
+    }
+    fn arg4(&self) -> usize {
+        self.a4 as usize
+    }
+    fn arg5(&self) -> usize {
+        self.a5 as usize
+    }
+    fn arg6(&self) -> usize {
+        self.a6 as usize
```

Comment:
> nit: Why are the arg registers usize, while the ret registers u64? I know they will end up being the same, but just wondering if it was chosen for a reason.

Params for both `argX()` and `retX()` methods are usize because it implements the [ArchState trait](https://github.com/rivosinc/hubris/blob/rivos/main-rv64/sys/kern/src/task.rs#L353). The `ArchState` trait uses usize because it is implemented by different architectures - 32 and 64bit.

For RV64, these registers are all fixed-width. Hence, `SavedState` uses `u64` and does the translation using `as usize` and `as u64` in the respective methods.

> It seems like a lot of the changes are switching from u32 to u64. Could we avoid diverging this file by using usize?

Do you mean diverging from riscv32.rs? I think its best to keep RV32 and RV64 support separate. There are going to be many more changes coming up that would probably take RV32 and RV64 further apart. We can always bring them back together into a common file if we realize later that they aren't really that different after all.

Reply:
Sounds good.